### PR TITLE
[AIMOD-291] Response level section experiments

### PR DESCRIPTION
**Pull request by mlplyler**
## References

JIRA: [AIMOD-291](https://mozilla-hub.atlassian.net/browse/AIMOD-291?atlOrigin=eyJpIjoiZjI4OTBkYTBmNjgzNDQyY2EzZmEyNTBiMjFiNmYwYmUiLCJwIjoiaiJ9)

## Description
Parses sections and looks for A/B pairs where A is the canonical id and B has a '__exp<name>' suffix. One type of section experiment is implemented: 5050. in the 5050 experiment, A or B is served in any given response at random. The canonical id is attached to version B (also already assigned to version A) so that block and follow is respected.

For non-US locales, there is a locale suffix attached to the id. The locale suffix must be a suffix and is last. We are changing section object construction such that instead of all __ suffix being dropped, only the __l locale suffix is dropped. The __exp suffix is dropped later. There is an additional hard drop of __ so that we keep the behavior of no __ surviving. 

Enum of assumptions:
* experimental sections must follow pattern <base>__exp<name>__l<locale>
* malformed experiments id are dropped. <base>__exp<name>__<anything>__l<locale > is dropped
* there can be only A/B versions for a base name. A/B/C is not supported and outcomes will be non-deterministic
* an experiment id without a matching base name is dropped. no solo experiment sections
* we always pass the base name to the user

## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ x] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [ x] The PR title starts with the JIRA issue reference, format example `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|skip|warn)]` keywords are applied to the last commit message (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [x ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

[AIMOD-291]: https://mozilla-hub.atlassian.net/browse/AIMOD-291?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: merino/curated_recommendations/corpus_backends/sections_backend.py [hunk 1]
```diff
@@ -132,9 +132,9 @@ async def fetch(self, surface_id: SurfaceId) -> list[CorpusSection]:
                 logger.info(f"Skipping inactive section {section['externalId']} for {surface_id}")
                 continue
 
-            # Strip any suffix (e.g., "__lEN_GB", "__lEN_CA") from externalId if present
-            # This handles locale suffixes and any future suffix patterns
-            external_id = section["externalId"].split("__")[0]
+            # Strip any locale suffix (e.g., "__lEN_GB", "__lEN_CA") from externalId if present.
```
Comment (rolf-moz) [inline merino/curated_recommendations/corpus_backends/sections_backend.py]: We are making the assumption here that experiment ID is always after the locale ID. 
It is probably good to use a utility function with a clear spec on supported and unsupported extension order. Maybe best to support both?

File: merino/curated_recommendations/corpus_backends/sections_backend.py [hunk 2]
```diff
@@ -132,9 +132,9 @@ async def fetch(self, surface_id: SurfaceId) -> list[CorpusSection]:
...
+            # Keep non-locale suffixes such as "__exp5050" intact for downstream processing.
+            external_id = section["externalId"].split("__l", 1)[0]
 
             section_obj = CorpusSection(
```
Comment (rolf-moz) [inline merino/curated_recommendations/corpus_backends/sections_backend.py:159]: Looks good but I wanted to suggest one possible improvement.

CorpusSection is an internal representation of a section. A different data type Section is returned to the Fx clients. Our previous google doc for the section ID in the new tab, we discussed having non-zero variant_id for a section indicating that it is a section variant.

Maybe we can start with implementing some of this now in Merino? Adding additional data structures in section_backend will keep repeat string comparisons happening every request.

Specifically, we could add to CorpusSection  an int field experiment_variant that defaults to 0 or None and store the original canonical section id. The alt section could be part of the list, or be linked as an attribute of CorpusSection (i.e. corpus section linking to an alternative slate). 

dedupe_experiment_sections could made much simpler with these changes. For example, it could go through the sections, and for a section that has a variant, decide which one to swap into the list.

In this method the section id would be normalized and you'd have two of the same section id in the list. Therefore you would never need to do a deep copy to create a new section during the request period.

File: merino/curated_recommendations/sections.py [hunk 3]
```diff
@@ -201,6 +199,76 @@ def _process_corpus_sections(
...
+        parsed_eid = clean_exp_id(eid)
+        if parsed_eid is None:
+            continue
+        can_eid, exp_type = parsed_eid
+        # Check if there is a matching non-experimental section.
+        if can_eid in id_to_section:
+            # Pick the winning section ID.
+            kept_id = resolve_section_experiment(can_eid, eid, exp_type)
+            # Replace the value while keeping the canonical ID.
+            id_to_result[can_eid] = deepcopy(id_to_section[kept_id])
```
Comment (rolf-moz) [inline merino/curated_recommendations/sections.py]: I think deep-copy will copy all the items individually and thus be computationally expensive. That would be bad... it's best to avoid deepcopy if possible... see my suggestion above. If you need to do a deep copy, maybe implement a custom function so you can be sure it does not recursively copy the list of items.

**Review by rolf-moz — CHANGES_REQUESTED** [review]

File: merino/curated_recommendations/corpus_backends/sections_backend.py (hunk 1 above)
Comment (mlplyler) [reply merino/curated_recommendations/corpus_backends/sections_backend.py]: We assume exp is BEFORE locale. 

Downstream from this line, we filter any id with exp that doesn't meet our specs. Are you suggestion we make this filter standalone and move to this location?

File: merino/curated_recommendations/sections.py (hunk 3 above)
Comment (mlplyler) [reply merino/curated_recommendations/sections.py]: Ah didn't realize the sections are cached. Let me try your suggestion.

File: merino/curated_recommendations/corpus_backends/sections_backend.py [hunk 4]
```diff
@@ -153,6 +170,28 @@ async def fetch(self, surface_id: SurfaceId) -> list[CorpusSection]:
                     for section_item in section["sectionItems"]
                 ],
             )
-            sections_list.append(section_obj)
+            parsed_sections.append(section_obj)
+
+        sections_list = []
```
Comment (rolf-moz) [inline merino/curated_recommendations/corpus_backends/sections_backend.py]: I see how we are a little more efficient this way, this code is called infrequently so we can maybe focus on readability

```
base_sections = [s for s in parsed_sections if s.experimentVariant == 0]
experimental_sections = [s for s in parsed_sections if s.experimentVariant != 0]

base_sections_by_id: dict[str, CorpusSection] = {
    s.externalId: s for s in base_sections
}

for section in experimental_sections:
    base = base_sections_by_id.get(section.externalId)
    if base and base.alternateSection is None:
        base.alternateSection = section

sections_list = base_sections
```

or if you can add a utility function for corpus sections.

This entire code block could be wrapped in a utility function also
```
base_sections = [s for s in parsed_sections if not s.is_experiment_section()]
experimental_sections = [s for s in parsed_sections if s.is_experiment_section()]

base_sections_by_id: dict[str, CorpusSection] = {
    s.externalId: s for s in base_sections
}

for section in experimental_sections:
    base = base_sections_by_id.get(section.externalId)
    if base and base.alternateSection is None:
        base.alternateSection = section

sections_list = base_sections
```

File: merino/curated_recommendations/corpus_backends/sections_backend.py
Comment (rolf-moz) [inline merino/curated_recommendations/corpus_backends/sections_backend.py:97]: Add comment here indicating experimental sections are removed from result and referenced separately

**Review by rolf-moz — APPROVED** [review]

**Comment by rolf-moz** [comment]
One side effect of the design that I suggested is that the legacy users (non sections) and mobile will get the base experiment without the variant. I don't know if there is enough traffic there to cause an issue.  The previous method would have likely merged both, increasing the number of items for the experimental sections.

It might be worth adding a to the integration tests for the legacy provider a case with an experimental section like this.

File: merino/curated_recommendations/corpus_backends/sections_backend.py (hunk 4 above)
Comment (mlplyler) [reply merino/curated_recommendations/corpus_backends/sections_backend.py]: ok took this thanks!

File: merino/curated_recommendations/corpus_backends/sections_backend.py [hunk 5]
```diff
@@ -35,6 +35,24 @@
...
+    idx = external_id.rfind(marker)
+    if idx <= 0:
+        return external_id.split("__", 1)[0], 0
+
+    base_id = external_id[:idx]
+    variant_id = external_id[idx + len(marker) :]
+    if variant_id.isdigit():
+        return base_id, int(variant_id)
+
+    return external_id.split("__", 1)[0], 0
```
Comment (mmiermans) [inline merino/curated_recommendations/corpus_backends/sections_backend.py:53]: Could you write unit tests against this function for any relevant edge case?

File: merino/curated_recommendations/corpus_backends/protocol.py [hunk 6]
```diff
@@ -95,6 +97,16 @@ class CorpusSection(BaseModel):
     iab: IABMetadata | None = None
     externalId: str
     createSource: CreateSource
+    experimentVariant: int = Field(
+        default=0,
+        exclude=True,
+        description="Internal experiment variant for the section. 0 means the canonical section.",
+    )
```
Comment (mmiermans) [inline merino/curated_recommendations/corpus_backends/protocol.py:104]: `exclude=True` will prevent this field from being sent to the client, right? How will clients report which branch they're in? Via the section name?

Comment (rolf-moz) [reply merino/curated_recommendations/corpus_backends/protocol.py:104]: At this point we don't have any reporting from the client. That is a client change that we'll do in the future. We know which items were in which variant, so that is how we'll measure the results.

Comment (rolf-moz) [reply merino/curated_recommendations/corpus_backends/protocol.py:104]: As far as i can tell the CorpusSection class is never directly serialized/unserialized in any way, so exclude=true simply indicates that the value is more of a transient data.

File: merino/curated_recommendations/corpus_backends/sections_backend.py (hunk 5 above)
Comment (mlplyler) [reply merino/curated_recommendations/corpus_backends/sections_backend.py:53]: Added TestParseSectionExternalId

File: merino/curated_recommendations/corpus_backends/protocol.py (hunk 6 above)
Comment (mmiermans) [reply merino/curated_recommendations/corpus_backends/protocol.py:104]: Gotcha. That means the A/B test is assumed to have distinct items?

**Review by mmiermans — APPROVED** [review]
